pkg/skaffold/build/knative/build.go

```diff
@@ -0,0 +1,158 @@
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/v1alpha2"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/util"
+	"github.com/knative/build/pkg/apis/build/v1alpha1"
+	build_v1alpha1 "github.com/knative/build/pkg/client/clientset/versioned/typed/build/v1alpha1"
+	"github.com/knative/build/pkg/logs"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+// BuildArtifact builds an artifact with knative Build CRD
+func BuildArtifact(ctx context.Context, out io.Writer, a *v1alpha2.Artifact, cfg *v1alpha2.KnativeBuild) (string, error) {
+	if err := testBuildCRD(); err != nil {
+		logrus.Errorln("Build CRD is not installed")
+		logrus.Errorln("Follow the installation guide: https://github.com/knative/build#getting-started")
+		logrus.Errorln("Usually, it's as simple as: kubectl create -f https://storage.googleapis.com/build-crd/latest/release.yaml")
+		return "", errors.Wrap(err, "Build CRD is not installed")
+	}
+
+	loadingRules := clientcmd.NewDefaultClientConfigLoadingRules()
+	kubeConfig := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(loadingRules, &clientcmd.ConfigOverrides{})
+	clientConfig, err := kubeConfig.ClientConfig()
+	if err != nil {
+		return "", errors.Wrap(err, "getting clientConfig")
+	}
+
+	deleteSecret, err := setupSecret(clientConfig, cfg)
+	if err != nil {
+		return "", errors.Wrap(err, "setting up secret")
+	}
+	defer deleteSecret()
+
+	deleteServiceAccount, err := setupServiceAccount(clientConfig, cfg)
+	if err != nil {
+		return "", errors.Wrap(err, "setting up service account")
+	}
+	defer deleteServiceAccount()
+
+	dockerfilePath := a.DockerArtifact.DockerfilePath
+
+	randomID := util.RandomID()
+	zipName := fmt.Sprintf("context-%s.tar.gz", randomID)
+	// TODO(dgageot): delete sources
+	if err := docker.UploadContextToGCS(ctx, a.Workspace, dockerfilePath, cfg.GCSBucket, zipName); err != nil {
+		return "", errors.Wrap(err, "uploading tar to gcs")
+	}
+
+	client, err := build_v1alpha1.NewForConfig(clientConfig)
+	if err != nil {
+		return "", errors.Wrap(err, "getting builds client")
+	}
+
+	imageDst := fmt.Sprintf("%s:%s", a.ImageName, randomID)
+	build := buildFor(imageDst, zipName, a, cfg)
+
+	builds := client.Builds(cfg.Namespace)
+	build, err = builds.Create(build)
+	if err != nil {
+		return "", errors.Wrap(err, "creating build")
+	}
+	defer func() {
+		if err := builds.Delete(build.Name, &metav1.DeleteOptions{}); err != nil {
+			logrus.Debugln("deleting build", err)
+		}
+	}()
+
+	if err := logs.Tail(ctx, out, build.Name, build.Namespace); err != nil {
+		return "", errors.Wrap(err, "streaming logs")
+	}
+
+	if err := waitForCompletion(builds, build); err != nil {
+		return "", errors.Wrap(err, "waiting for completion")
+	}
+
+	return imageDst, nil
+}
+
+func waitForCompletion(builds build_v1alpha1.BuildInterface, build *v1alpha1.Build) error {
+	for {
```

Comment:
why busy wait?

Reply:
@AkihiroSuda I'm not sure I understand your comment

Reply:
Does not this loop consume CPU too much?

Reply:
No, it's done just after the logs are done tailing so it usually succeeds i getting the status the first time. But since there's a few ms between the logs being done and the status being updated, it sometimes requires a retry.